Add Ruff formatter and pre-commit configurations

Zeitsperre opened the pull request:
# Changes

* Added a .pre-commit configuration
* Added configurations for `ruff`
* Ran safe and aggressive fixes over the codebase
* Add .idea folder to the `.gitignore`

Zeitsperre committed:
add ruff formatter, poetry-based versioning

Zeitsperre committed:
run pre-commit linters with safe fixes

Zeitsperre committed:
add .idea to .gitignore

Zeitsperre committed:
run pre-commit linters with aggressive fixes

Zeitsperre committed:
remove poetry-version-plugin configuration

bneijt commented on .github/workflows/main.yml:
```diff
@@ -16,4 +16,4 @@ jobs:
       - run: pip install poetry==1.1.11
       - run: poetry install
       - run: poetry run black --check .
```
moving to ruff, we should remove black in favor of ruff-format.

Makes sense to me to just run pre-commit check here.

bneijt commented on .github/workflows/main.yml:
```diff
@@ -16,4 +16,4 @@ jobs:
       - run: pip install poetry==1.1.11
```
Really old version of poetry, probably why the pipeline PR pipeline is failing, let's bump this.

bneijt commented:
Thank you for your great work and sorry for not getting back earlier. All changes seem sane to me, but pr pipeline is failing. Probably because it's on an old poetry version. I think if you fix those changes we could merge this asap.

Zeitsperre committed:
update testing and publishing workflows

Zeitsperre committed:
adjust lowest pre-commit version to include Python3.8, regenerate poetry lock file with `--no-update`

Zeitsperre commented:
@bneijt 

I've addressed your comments here. There are definitely some overlapping changes with #19, but I'll deal with those once this is merged. I've added a linting step to the `main.yml` so that if the checks don't pass linting, the rest of the tests won't run.

I've also updated the poetry version and regenerated the lock file with `--no-update` so that changes aren't as destructive.

The `ruff` formatter is reporting around 71 errors, mostly from bare exceptions and some undefined variables. I would suggest running the tests locally to ensure it looks good on your end, merging this Pull Request (overriding changes), then addressing the problems (or ignoring them) in #19.

All the best!

Zeitsperre committed:
add ruff ignore codes

Zeitsperre commented:
@bneijt 

I've added some `ignore` codes to the `ruff` configuration so that we can at the very least merge PR this without issue. The rest of the fixes are about ready in my other PR.
